File: pkg/cli/format.go
// Package cli formats system monitoring data for terminal output.
package cli

import (
	"fmt"
	"io"
	"strings"
)

const (
	nvmeDataUnitBytes = 512_000
)

// formatBytes converts a byte count to a human-readable string.
func formatBytes(n uint64) string {
	const (
		kb = 1024
		mb = kb * 1024
		gb = mb * 1024
		tb = gb * 1024
	)
	switch {
	case n == 0:
		return "0 B"
	case n >= tb:
		return fmt.Sprintf("%.1f TB", float64(n)/float64(tb))
	case n >= gb:
		return fmt.Sprintf("%.1f GB", float64(n)/float64(gb))
	case n >= mb:
		return fmt.Sprintf("%.1f MB", float64(n)/float64(mb))
	case n >= kb:
		return fmt.Sprintf("%.1f KB", float64(n)/float64(kb))
	default:
		return fmt.Sprintf("%d B", n)
	}
}

// formatDuration converts a duration in seconds to a human-readable string.
func formatDuration(seconds uint64) string {
	const (
		minute = 60
		hour   = minute * 60
		day    = hour * 24
	)
	days := seconds / day
	hours := (seconds % day) / hour
	mins := (seconds % hour) / minute

	if days > 0 {
		return fmt.Sprintf("%dd %dh %dm", days, hours, mins)
	}
	if hours > 0 {
		return fmt.Sprintf("%dh %dm", hours, mins)
	}
	return fmt.Sprintf("%dm", mins)
}

// formatPercent formats a float64 as a percentage string.
func formatPercent(v float64) string {
	return fmt.Sprintf("%.1f%%", v)
}

// formatBar renders an ASCII progress bar of the given width.
// filled blocks use '█', empty blocks use '░'.
func formatBar(percent float64, width int) string {
	if width <= 0 {
		return ""
	}
	if percent < 0 {
		percent = 0
	}
	if percent > 100 {
		percent = 100
	}
	filled := int(percent/100.0*float64(width) + 0.5)
	if filled > width {
		filled = width
	}
	return strings.Repeat("█", filled) + strings.Repeat("░", width-filled)
}

// formatHours converts hours to a human-readable years/days string.
func formatHours(hours uint64) string {
	const (
		hoursPerDay  = 24
		daysPerYear  = 365
		hoursPerYear = hoursPerDay * daysPerYear
	)
	if hours == 0 {
		return "0 hours"
	}
	years := hours / hoursPerYear
	days := (hours % hoursPerYear) / hoursPerDay
	if years > 0 {
		return fmt.Sprintf("%d years, %d days", years, days)
	}
	return fmt.Sprintf("%d days", days)
}

// formatDataUnits converts NVMe data units (each = 512,000 bytes) to a
// human-readable byte string.
func formatDataUnits(units uint64) string {
	return formatBytes(units * nvmeDataUnitBytes)
}

// formatTable writes a simple column-aligned table to w.
// Column widths are determined by the widest value in each column.
func formatTable(w io.Writer, headers []string, rows [][]string) {
	widths := make([]int, len(headers))
	for i, h := range headers {
		widths[i] = len(h)
	}
	for _, row := range rows {
		for i, cell := range row {
			if i < len(widths) && len(cell) > widths[i] {
				widths[i] = len(cell)
			}
		}
	}

	// header row
	for i, h := range headers {
		if i > 0 {
			fmt.Fprint(w, "  ")
		}
		fmt.Fprintf(w, "%-*s", widths[i], h)
	}
	fmt.Fprintln(w)

	// separator
	for i, w2 := range widths {
		if i > 0 {
			fmt.Fprint(w, "  ")
		}
		fmt.Fprint(w, strings.Repeat("-", w2))
	}
	fmt.Fprintln(w)

	// data rows
	for _, row := range rows {
		for i, cell := range row {
			if i >= len(widths) {
				break
			}
			if i > 0 {
				fmt.Fprint(w, "  ")
			}
			fmt.Fprintf(w, "%-*s", widths[i], cell)
		}
		fmt.Fprintln(w)
	}
}

File: cmd/sysmon/frontend/src/lib/format.ts
// Human-readable formatting utilities.

const BYTE_UNITS = ["B", "KB", "MB", "GB", "TB", "PB"] as const;

/**
 * formatBytes converts a raw byte count to the most readable unit string.
 * e.g. 1536 -> "1.5 KB"
 */
export function formatBytes(bytes: number, decimals = 1): string {
  if (bytes === 0) return "0 B";
  const k = 1024;
  const i = Math.min(
    Math.floor(Math.log(bytes) / Math.log(k)),
    BYTE_UNITS.length - 1
  );
  const value = bytes / Math.pow(k, i);
  return `${value.toFixed(decimals)} ${BYTE_UNITS[i]}`;
}

/**
 * formatBytesRate converts a bytes-per-second rate to a human-readable string.
 * e.g. 2048 -> "2.0 KB/s"
 */
export function formatBytesRate(bytesPerSec: number, decimals = 1): string {
  return `${formatBytes(bytesPerSec, decimals)}/s`;
}

/**
 * formatDuration converts a duration in seconds to a human-readable string.
 * e.g. 90061 -> "1d 1h 1m"
 */
export function formatDuration(seconds: number): string {
  if (seconds <= 0) return "0s";

  const days = Math.floor(seconds / 86400);
  const hours = Math.floor((seconds % 86400) / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);

  const parts: string[] = [];
  if (days > 0) parts.push(`${days}d`);
  if (hours > 0) parts.push(`${hours}h`);
  if (minutes > 0) parts.push(`${minutes}m`);
  if (parts.length === 0) parts.push(`${seconds}s`);

  return parts.join(" ");
}

/**
 * formatPercent formats a floating-point percentage with one decimal place.
 */
export function formatPercent(value: number): string {
  return `${value.toFixed(1)}%`;
}

/**
 * formatMHz formats a CPU frequency.
 */
export function formatMHz(mhz: number): string {
  if (mhz >= 1000) {
    return `${(mhz / 1000).toFixed(2)} GHz`;
  }
  return `${mhz.toFixed(0)} MHz`;
}

/**
 * usageColor returns a DaisyUI semantic color class based on a usage percentage.
 * green < 50%, yellow 50-80%, red > 80%
 */
export function usageColor(percent: number): string {
  if (percent >= 80) return "text-error";
  if (percent >= 50) return "text-warning";
  return "text-success";
}

/**
 * usageBgColor returns a progress bar color class for a given usage percentage.
 */
export function usageProgressColor(percent: number): string {
  if (percent >= 80) return "progress-error";
  if (percent >= 50) return "progress-warning";
  return "progress-success";
}

/**
 * usageEchartsColor returns an ECharts color string for a usage percentage.
 */
export function usageEchartsColor(percent: number): string {
  if (percent >= 80) return "#f87272"; // DaisyUI error red
  if (percent >= 50) return "#fbbd23"; // DaisyUI warning yellow
  return "#36d399"; // DaisyUI success green
}

/**
 * formatHoursHuman converts a duration in hours to a human-readable string.
 * e.g. 26280 (3 years) -> "3 years, 0 days"
 * e.g. 400 -> "16 days, 16 hours"
 */
export function formatHoursHuman(hours: number): string {
  if (hours <= 0) return "0 hours";

  const HOURS_PER_YEAR = 8760;
  const HOURS_PER_DAY = 24;

  if (hours >= HOURS_PER_YEAR) {
    const years = Math.floor(hours / HOURS_PER_YEAR);
    const remainingDays = Math.floor((hours % HOURS_PER_YEAR) / HOURS_PER_DAY);
    if (remainingDays === 0) return `${years} year${years !== 1 ? "s" : ""}`;
    return `${years} year${years !== 1 ? "s" : ""}, ${remainingDays} day${remainingDays !== 1 ? "s" : ""}`;
  }

  if (hours >= HOURS_PER_DAY) {
    const days = Math.floor(hours / HOURS_PER_DAY);
    const remainingHours = Math.floor(hours % HOURS_PER_DAY);
    if (remainingHours === 0) return `${days} day${days !== 1 ? "s" : ""}`;
    return `${days} day${days !== 1 ? "s" : ""}, ${remainingHours} hour${remainingHours !== 1 ? "s" : ""}`;
  }

  const h = Math.floor(hours);
  return `${h} hour${h !== 1 ? "s" : ""}`;
}

/**
 * formatDataUnits converts NVMe data units to a human-readable byte string.
 * Each NVMe data unit = 512,000 bytes (1000 sectors of 512 bytes).
 * e.g. 85070 -> "43.5 TB"
 */
export function formatDataUnits(units: number): string {
  return formatBytes(units * 512000);
}

/**
 * formatWatts formats a power value in watts to a human-readable string.
 * e.g. 440.5 -> "440.5 W"
 */
export function formatWatts(w: number): string {
  return `${w.toFixed(1)} W`;
}

/**
 * formatTemp formats a temperature in Celsius to a human-readable string.
 * Returns "-" when the value is 0 (not reported).
 * e.g. 52.0 -> "52.0°C"
 */
export function formatTemp(celsius: number): string {
  if (celsius === 0) return "-";
  return `${celsius.toFixed(1)}°C`;
}

/**
 * tempColor returns a DaisyUI semantic color class based on proximity to the
 * critical threshold. green < 70%, yellow 70-85%, red > 85% of crit.
 */
export function tempColor(celsius: number, critCelsius: number): string {
  if (critCelsius <= 0) {
    // No threshold data: use fixed bands.
    if (celsius >= 90) return "text-error";
    if (celsius >= 70) return "text-warning";
    return "text-success";
  }
  const ratio = celsius / critCelsius;
  if (ratio >= 0.85) return "text-error";
  if (ratio >= 0.70) return "text-warning";
  return "text-success";
}

/**
 * tempProgressColor returns a DaisyUI progress bar color class based on
 * proximity to the critical threshold.
 */
export function tempProgressColor(celsius: number, critCelsius: number): string {
  if (critCelsius <= 0) {
    if (celsius >= 90) return "progress-error";
    if (celsius >= 70) return "progress-warning";
    return "progress-success";
  }
  const ratio = celsius / critCelsius;
  if (ratio >= 0.85) return "progress-error";
  if (ratio >= 0.70) return "progress-warning";
  return "progress-success";
}

/**
 * formatCores renders a "percent of one core" value as a core count.
 * 1703 becomes "17.0 cores", which is what the number actually means.
 */
export function formatCores(percentOfOneCore: number): string {
  const cores = percentOfOneCore / 100;
  if (cores < 10) return `${cores.toFixed(1)} cores`;
  return `${Math.round(cores)} cores`;
}

/**
 * percentOfCapacity converts a "percent of one core" value into a share of the
 * capacity available to that workload, so it lands on the 0-100 scale people
 * expect. capacityCores is the vCPU count for a guest, the quota for a limited
 * container, or the host thread count when nothing caps it.
 */
export function percentOfCapacity(
  percentOfOneCore: number,
  capacityCores: number
): number {
  if (capacityCores <= 0) return 0;
  return percentOfOneCore / capacityCores;
}
